File: solutions/cpp/robot-simulator/2/robot_simulator.cpp
#include "robot_simulator.h"
#include <functional>
#include <map>

namespace robot_simulator {

void bump(Bearing& b, int n) {
    constexpr unsigned limit = static_cast<unsigned>(Bearing::COUNT);
    b = static_cast<Bearing>((static_cast<unsigned>(b) + n) % limit);
}

Robot::Robot(std::pair<int, int> position, Bearing bearing): 
    position(position), bearing(bearing) {
}

void Robot::turn_right() {
    bump(bearing, 1);
}

void Robot::turn_left() {
    bump(bearing, -1);
}

void Robot::advance() {
    static const std::map<Bearing, std::pair<int, int>> offsets_for_bearing = {
        {Bearing::NORTH, {0, 1}},
        {Bearing::EAST, {1, 0}},
        {Bearing::SOUTH, {0, -1}},
        {Bearing::WEST, {-1, 0}}
    };

    position.first += offsets_for_bearing.at(bearing).first;    
    position.second += offsets_for_bearing.at(bearing).second;
}    
    
void Robot::execute_sequence(const std::string& commands) {
    std::map<char, std::function<void()> > action_for_command {
        {'L', std::bind(&Robot::turn_left, this)}, 
        {'R', std::bind(&Robot::turn_right, this)}, 
        {'A', std::bind(&Robot::advance, this)}
    };
    for (auto c: commands)
        action_for_command.at(c)();
}

}  // namespace robot_simulator

File: solutions/clojure/robot-simulator/2/src/robot_simulator.clj
(ns robot-simulator)

(defn robot
  [position bearing]
  {:bearing bearing :coordinates position})

(def bearings
  [:north :east :south :west])

(defn offset-for-bearing [bearing]
  (case bearing
    :north {:dx  0  :dy  1}
    :east  {:dx  1  :dy  0}
    :south {:dx  0  :dy -1}
    :west  {:dx -1  :dy  0}))

(defn bump-circular [direction limit value]
  (let [next-value (+ direction value)]
    (cond
      (< next-value 0) (- limit 1)
      (>= next-value limit) 0
      :else next-value)))
      
(defn bump-bearing [direction bearing]
  (let [index (.indexOf bearings bearing)
        limit (count bearings)]
    (get bearings (bump-circular direction limit index))))

(defn translate [{:keys [dx dy]} {:keys [x y]}]
  {:x (+ x dx) :y (+ y dy)})
  
(defn rotate [direction robot]
  (update robot :bearing (partial bump-bearing direction)))

(defn advance [robot]
  (let [offset (offset-for-bearing (:bearing robot))]
    (update robot :coordinates (partial translate offset))))

(defn execute
  [command robot]
  (case command
    \L (rotate -1 robot)
    \R (rotate 1 robot)
    \A (advance robot)))

(defn simulate
  [commands robot]
  (if (= commands "") robot 
    (recur (subs commands 1) (execute (get commands 0) robot))))
